File: lc_go/lcGo/binarySearch/findClosestElements.go
package main

import (
    "container/heap"
    "fmt"
    "math"
    "sort"
)

// 方法1: 二分扩展法
func findClosestElements(arr []int, k int, x int) []int {
    n := len(arr)
    idx := sort.SearchInts(arr, x)
    
    left, right := idx-1, idx
    
    for right-left-1 < k {
        if left < 0 {
            right++
        } else if right >= n {
            left--
        } else {
            if x-arr[left] <= arr[right]-x {
                left--
            } else {
                right++
            }
        }
    }
    
    return arr[left+1 : right]
}

// 方法2: 滑动窗口法
func findClosestElementsSlidingWindow(arr []int, k int, x int) []int {
    left, right := 0, len(arr)-k
    
    for left < right {
        mid := left + (right-left)/2
        if x-arr[mid] > arr[mid+k]-x {
            left = mid + 1
        } else {
            right = mid
        }
    }
    
    return arr[left : left+k]
}

// 大根堆实现
type MaxHeap [][]int // [distance, value, index]

func (h MaxHeap) Len() int { return len(h) }
func (h MaxHeap) Less(i, j int) bool {
    if h[i][0] != h[j][0] {
        return h[i][0] > h[j][0] // 距离大的在前
    }
    return h[i][1] > h[j][1] // 距离相同时值大的在前
}
func (h MaxHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }
func (h *MaxHeap) Push(x interface{}) {
    *h = append(*h, x.([]int))
}
func (h *MaxHeap) Pop() interface{} {
    old := *h
    n := len(old)
    x := old[n-1]
    *h = old[0 : n-1]
    return x
}

// 方法3: 大根堆法
func findClosestElementsMaxHeap(arr []int, k int, x int) []int {
    h := &MaxHeap{}
    heap.Init(h)
    
    for i, num := range arr {
        distance := int(math.Abs(float64(num - x)))
        if h.Len() < k {
            heap.Push(h, []int{distance, num, i})
        } else {
            top := (*h)[0]
            if distance < top[0] || (distance == top[0] && num < top[1]) {
                heap.Pop(h)
                heap.Push(h, []int{distance, num, i})
            }
        }
    }
    
    // 提取结果并排序
    type pair struct {
        index int
        value int
    }
    
    var temp []pair
    for h.Len() > 0 {
        item := heap.Pop(h).([]int)
        temp = append(temp, pair{item[2], item[1]})
    }
    
    sort.Slice(temp, func(i, j int) bool {
        return temp[i].index < temp[j].index
    })
    
    result := make([]int, len(temp))
    for i, p := range temp {
        result[i] = p.value
    }
    return result
}

// 方法3优化: 大根堆法 - 提前终止
func findClosestElementsOptimizedHeap(arr []int, k int, x int) []int {
    h := &MaxHeap{}
    heap.Init(h)
    
    for i, num := range arr {
        distance := int(math.Abs(float64(num - x)))
        if h.Len() < k {
            heap.Push(h, []int{distance, num, i})
        } else {
            top := (*h)[0]
            if distance < top[0] || (distance == top[0] && num < top[1]) {
                heap.Pop(h)
                heap.Push(h, []int{distance, num, i})
            } else if distance > top[0] {
                break // 提前终止
            }
        }
    }
    
    type pair struct {
        index int
        value int
    }
    
    var temp []pair
    for h.Len() > 0 {
        item := heap.Pop(h).([]int)
        temp = append(temp, pair{item[2], item[1]})
    }
    
    sort.Slice(temp, func(i, j int) bool {
        return temp[i].index < temp[j].index
    })
    
    result := make([]int, len(temp))
    for i, p := range temp {
        result[i] = p.value
    }
    return result
}

func main() {
    arr := []int{1, 2, 3, 4, 5}
    k, x := 4, 3
    
    fmt.Println("二分扩展法:", findClosestElements(arr, k, x))
    fmt.Println("滑动窗口法:", findClosestElementsSlidingWindow(arr, k, x))
    fmt.Println("大根堆法:", findClosestElementsMaxHeap(arr, k, x))
    fmt.Println("优化堆法:", findClosestElementsOptimizedHeap(arr, k, x))
}

File: lc_ts/lcTs/binarysearch/findClosestElements.ts
class Solution {
    // 方法1: 二分扩展法
    findClosestElements(arr: number[], k: number, x: number): number[] {
        const n = arr.length;
        let idx = this.binarySearch(arr, x);
        
        let left = idx - 1, right = idx;
        
        while (right - left - 1 < k) {
            if (left < 0) {
                right++;
            } else if (right >= n) {
                left--;
            } else {
                if (x - arr[left] <= arr[right] - x) {
                    left--;
                } else {
                    right++;
                }
            }
        }
        
        return arr.slice(left + 1, right);
    }
    
    // 方法2: 滑动窗口法
    findClosestElementsSlidingWindow(arr: number[], k: number, x: number): number[] {
        let left = 0, right = arr.length - k;
        
        while (left < right) {
            const mid = Math.floor((left + right) / 2);
            if (x - arr[mid] > arr[mid + k] - x) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        
        return arr.slice(left, left + k);
    }
    
    // 方法3: 大根堆法
    findClosestElementsMaxHeap(arr: number[], k: number, x: number): number[] {
        const heap = new MaxHeap<[number, number, number]>((a, b) => {
            if (a[0] !== b[0]) return a[0] - b[0];
            return a[1] - b[1];
        });
        
        for (let i = 0; i < arr.length; i++) {
            const distance = Math.abs(arr[i] - x);
            if (heap.size() < k) {
                heap.push([distance, arr[i], i]);
            } else {
                const top = heap.peek();
                if (top && (distance < top[0] || (distance === top[0] && arr[i] < top[1]))) {
                    heap.pop();
                    heap.push([distance, arr[i], i]);
                }
            }
        }
        
        const temp: [number, number][] = [];
        while (heap.size() > 0) {
            const item = heap.pop()!;
            temp.push([item[2], item[1]]);
        }
        
        temp.sort((a, b) => a[0] - b[0]);
        return temp.map(item => item[1]);
    }
    
    private binarySearch(arr: number[], target: number): number {
        let left = 0, right = arr.length;
        while (left < right) {
            const mid = Math.floor((left + right) / 2);
            if (arr[mid] < target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }
}

// 泛型大根堆实现
class MaxHeap<T> {
    private heap: T[] = [];
    private compare: (a: T, b: T) => number;
    
    constructor(compareFn: (a: T, b: T) => number) {
        this.compare = compareFn;
    }
    
    push(val: T): void {
        this.heap.push(val);
        this.bubbleUp(this.heap.length - 1);
    }
    
    pop(): T | null {
        if (this.heap.length === 0) return null;
        if (this.heap.length === 1) return this.heap.pop()!;
        
        const top = this.heap[0];
        this.heap[0] = this.heap.pop()!;
        this.bubbleDown(0);
        return top;
    }
    
    peek(): T | null {
        return this.heap[0] || null;
    }
    
    size(): number {
        return this.heap.length;
    }
    
    private bubbleUp(index: number): void {
        while (index > 0) {
            const parentIndex = Math.floor((index - 1) / 2);
            if (this.compare(this.heap[index], this.heap[parentIndex]) <= 0) break;
            [this.heap[index], this.heap[parentIndex]] = [this.heap[parentIndex], this.heap[index]];
            index = parentIndex;
        }
    }
    
    private bubbleDown(index: number): void {
        while (true) {
            let largest = index;
            const leftChild = 2 * index + 1;
            const rightChild = 2 * index + 2;
            
            if (leftChild < this.heap.length && 
                this.compare(this.heap[leftChild], this.heap[largest]) > 0) {
                largest = leftChild;
            }
            
            if (rightChild < this.heap.length && 
                this.compare(this.heap[rightChild], this.heap[largest]) > 0) {
                largest = rightChild;
            }
            
            if (largest === index) break;
            [this.heap[index], this.heap[largest]] = [this.heap[largest], this.heap[index]];
            index = largest;
        }
    }
}

// 函数式编程解法
const findClosestElementsFunctional = (arr: number[], k: number, x: number): number[] => {
    // 计算所有元素到x的距离
    const distances = arr.map((num, index) => ({
        value: num,
        distance: Math.abs(num - x),
        index
    }));
    
    // 排序并取前k个
    distances.sort((a, b) => {
        if (a.distance !== b.distance) return a.distance - b.distance;
        return a.value - b.value;
    });
    
    // 按原数组顺序返回
    const result = distances.slice(0, k);
    result.sort((a, b) => a.index - b.index);
    return result.map(item => item.value);
};

// 测试
const sol = new Solution();
const arr = [1, 2, 3, 4, 5];
const k = 4, x = 3;

console.log("二分扩展法:", sol.findClosestElements(arr, k, x));
console.log("滑动窗口法:", sol.findClosestElementsSlidingWindow(arr, k, x));
console.log("大根堆法:", sol.findClosestElementsMaxHeap(arr, k, x));
console.log("函数式编程:", findClosestElementsFunctional(arr, k, x));